### PR TITLE
Feature gst checkpointing

### PR DESCRIPTION
**Pull request by coreyostrove**
This PR gives an initial implementation for checkpointing iterative GST fits and introduces some basic and extensible infrastructure for checkpointing protocols more broadly going forward. For this initial implementation there is currently only full support for checkpointing `ModelTest`, ` GateSetTomography` and 'StandardGST' protocols. Some highlights:

1. Adds ProtocolCheckpoint, GateSetTomographyCheckpoint, ModelTestCheckpoint and StandardGSTCheckpoint objects, each of which can be serialized natively to json.
2. Protocol run methods (and related driver functions) now take as input kwargs allowing one to give a previously generated checkpoint object if one exists, a path to write checkpoint files to during the course of the fit and finally an option to disable checkpointing entirely (checkpointing is on by default).

This should hopefully reduce one of the common pain points users encounter wherein their fits fail for some mundane reason (like running out of memory or cluster allocation time etc.) and then they need to restart an expensive calculation from scratch.

I have added some unit tests (these currently live in the test_drivers module in test_packages for convenience), which all seem to be passing at present. We do have some tests failing in the extras, but these look to be unrelated to new code beyond the fact that the new code made these bugs visible (discussed further in #346). So given that I'd advocate for merging this in despite those failing tests. Ultimately we'll want to disable checkpointing in these particular failing unit tests---I had done a pass through the tests to disable most of these to avoid needing to deal with file hygiene on the runners for tests that aren't related to checkpointing but missed these by accident---but for now I'd argue is useful to keep it enabled since that is what made these serialization bugs apparent.

Unrelatedly this includes a deprecation of the comma-separated string syntax for modes in StandardGST and some fixes to unit tests to account for that. 

Feedback and questions welcome.

### CODE REVIEW

File: pygsti/algorithms/core.py
Comment (sserita) [inline pygsti/algorithms/core.py:1]: Is it worth refactoring this now to avoid duplicating the iterative_gst function?

**Review by sserita — COMMENTED** [review]
This all looks really good. I appreciate both your commitment to updating examples and tests to avoid deprecation warnings and the new unit tests. Only have one comment for us to touch base on before approving.

File: pygsti/algorithms/core.py
Comment (coreyostrove) [reply pygsti/algorithms/core.py:1]: Ah, good catch. It had slipped my mind that we had discussed merging these different code paths together. I'll refactor the run_iterative_gst function to use the generator version under the hood and push that change.

Comment (coreyostrove) [reply pygsti/algorithms/core.py:1]: Refactor done and I've also added a new test targeting a new option specific to use of the generator version. Thanks again for the feedback.

**Review by sserita — APPROVED** [review]
Thanks Corey! Looks great.

**Pull request merged**